Doc for Dynamic and Static routing (#419)
apache/cloudstack-documentation: source/adminguide/networking/dynamic_static_routing.rst

- ROUTED. For isolated networks in ROUTED mode, the VR no longer supports Source NAT, Static NAT, Load Balancer, Port Forwarding and Vpn. The supported services are Dns, Dhcp, Userdata, Firewall (for isolated networks) and Network ACL (for vpc and vpc networks).

About Routing mode
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Routing mode indicates how routing will operate with the isolated networks with ROUTED network mode.
There are two valid options

- Static. The operators need to add the static routes to the isolated networks or VPCs in the upstream router manually.
- Dynamic. The AS number will be automatically allocated, and BGP peer sessions will be set up automatically in the VR of the isolated networks or VPCs. The operators need to add the AS number ranges and BGP peers for each zone before creating network with Dynamic routing mode.

Manage IPv4 Subnets for Zone
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Like IPv6 prefixes, operators need to configure the IPv4 subnets for zone, which will be eventually used by guest networks.

> Supported CloudStack API for operators to manage the IPv4 subnets for zone are:

Reviewer: ```suggestion
Supported CloudStack APIs for operators to manage the IPv4 subnets for zone are:
```
Not sure if we need to specify CloudStack each section

- **dedicateIpv4SubnetForZone** : dedicate an IPv4 subnet for zone to a domain or an account
- **deleteIpv4SubnetForZone** : delete an IPv4 subnet for zone
- **listIpv4SubnetsForZone** : list IPv4 subnets for zone
- **releaseIpv4SubnetForZone** : release a dedicated IPv4 subnet for zone from a domain or an account
- **updateIpv4SubnetForZone** : update an IPv4 subnet for zone

Operators (root admins) can manage the IPv4 subnets for zone by navigating to Infrastructure -> Zones -> IPv4 Subnets
|manage-ipv4-subnets-for-zone.png|

Manage IPv4 Subnets for Guest Networks
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Unlike IPv6 (each isolated network with IPv6 support gets a /64 IPv6 network), operators need to manage IPv4 subnets for guest networks.
An IPv4 subnet for guest networks is created from its parent which is a IPv4 subnet for zone.

There are some global settings which can be set for each account. See below

.. cssclass:: table-striped table-bordered table-hover

================================================= ========================
Configuration                                       Description
================================================= ========================

> routed.ipv4.network.cidr.auto.allocation.enabled    whether the auto-allocation of network CIDR for routed network is enabled or not. True by default.

Reviewer: ```suggestion
routed.ipv4.network.cidr.auto.allocation.enabled    Whether the auto-allocation of network CIDR for routed network is enabled or not. True by default.
```

routed.ipv4.network.max.cidr.size                   The maximum value of the cidr size for isolated networks in ROUTED mode	
routed.ipv4.network.min.cidr.size                   The minimum value of the cidr size for isolated networks in ROUTED mode	
routed.ipv4.vpc.max.cidr.size                       The maximum value of the cidr size for VPC in ROUTED mode	
routed.ipv4.vpc.min.cidr.size                       The minimum value of the cidr size for VPC in ROUTED mode	
================================================= ========================

> Supported CloudStack API for operators to manage the IPv4 subnets for guest networks are:

Author: ```suggestion
Supported CloudStack APIs for operators to manage the IPv4 subnets for guest networks are:
```

If cidrsize is specified, CloudStack will allocate an IPv4 subnet for guest network to the net network

- Check if there is an IPv4 subnet with same CIDR size available,
- If not, and setting "routed.ipv4.network.cidr.auto.allocation.enabled" is true for account, allocate an IPv4 subnet for the new network, from the IPv4 subnet for zone which the account can access.
- Otherwise, the network creation fails.

When the network is implemented, the Ipv4 routes are displayed in the network details page.

|routed-ipv4-routes.png|

Create Network with Static Routing for IPv6
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The IPv6 static routing has been introduced in Apache CloudStack 4.17.0.0.
For more information, see `“IPv6 support for isolated networks and VPC Network Tiers” <../plugins/ipv6.html#isolated-network-and-vpc-network-tier>`_.

Users can create network with static routing for both IPv4 and IPv6, if the network offering supports DualStack.

Manage IPv4 Routing Firewall
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Users can manage the IPv4 Routing firewalls by navigating to Network -> Guest Networks -> choose a network -> IPv4 Routing Firewall

|routed-ipv4-routing-firewall.png|

> Supported CloudStack API for operators to manage the IPv4 Routing firewall rules are:

Author: ```suggestion
Supported CloudStack APIs for operators to manage the IPv4 Routing firewall rules are:
```

- **createRoutingFirewallRule** : create an IPv4 routing firewall rule
- **updateRoutingFirewallRule** : update an IPv4 routing firewall rule
- **deleteRoutingFirewallRule** : delete an IPv4 routing firewall rule
- **listRoutingFirewallRules** : list IPv4 routing firewall rules

Manage AS number for Dynamic Routing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To create network with dynamic routing, operators must add AS number ranges in advance by navigating to Infrastructure -> Zones -> choose a zone -> AS Number.

|dynamic-routing-as-number-ranges.png|

> Supported CloudStack API for operators to manage the AS number ranges and AS numbers are:

Author: ```suggestion
Supported CloudStack APIs for operators to manage the AS number ranges and AS numbers are:
```

- **createASNRange** : Creates a range of Autonomous Systems for BGP Dynamic Routing
- **listASNRanges** : List Autonomous Systems Number Ranges
- **deleteASNRange** : deletes a range of Autonomous Systems for BGP Dynamic Routing
- **listASNumbers** : List Autonomous Systems Numbers
- **releaseASNumber** : Releases an AS Number back to the pool

Operators can list the AS numbers by navigating to Network -> AS Numbers

|dynamic-routing-as-numbers.png|

Manage BPG peers for Dynamic Routing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To create network with dynamic routing, operators must add BGP peers in advance. Guest networks with Dynamic Routing will connect to all BGP peers the account can access.

|dynamic-routing-bgp-peers.png|

> Supported CloudStack API for operators to manage the BGP peers are:

Author: ```suggestion
Supported CloudStack APIs for operators to manage the BGP peers are:
```

- **createBgpPeer** : create a BGP peer
- **dedicateBgpPeer** : dedicate a BGP peer to a domain or an account
- **deleteBgpPeer** : delete a BGP peer
- **listBgpPeers** : list BGP peers

> - **releaseBgpPeer** : release a dedicated IPv4 subnet for zone from a domain or an account

Author: ```suggestion
- **releaseBgpPeer** : release a dedicated BGP peer from a domain or an account
```

.. Licensed to the Apache Software Foundation (ASF) under one
   or more contributor license agreements.  See the NOTICE file
   distributed with this work for additional information#
   regarding copyright ownership.  The ASF licenses this file
   to you under the Apache License, Version 2.0 (the
   "License"); you may not use this file except in compliance
   with the License.  You may obtain a copy of the License at
   http://www.apache.org/licenses/LICENSE-2.0
   Unless required by applicable law or agreed to in writing,
   software distributed under the License is distributed on an
   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
   KIND, either express or implied.  See the License for the
   specific language governing permissions and limitations
   under the License.

Dynamic and Static Routing
-----------------------------

For VMs on Isolated networks, the IP of VMs are not publicly accessible.

> To access the VMs from the Internet, Users need to create Load balanceing rules,

Reviewer 2: Load balancing  spelling

During the network creation, CloudStack will

- Allocate an AS number to the network
- If the network owner does not have dedicated BGP peers, or account setting "use.system.bgp.peers" is set to true, configure BGP sessions in the network VR to connect to all BGP peers the network owner can access.
- If the network owner has dedicated BGP peers, and account setting "use.system.bgp.peers" is set to false, configure BGP sessions in the network VR to connect to all dedicated BGP peers of the domain and the network owner.

ROOT admin can change BGP peers of an existing network with Dynamic routing. After that, the network VR will only connect to selected BGP peers.

|dynamic-routing-change-network-bgp-peers.png|

Create VPC with Dynamic Routing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The creation of VPC with Dynamic routing is almost as VPC with static routing. CloudStack will allocate an AS number to the VPC, and 
- If the VPC owner does not have dedicated BGP peers, or account setting "use.system.bgp.peers" is set to true, configure BGP sessions in the VPC VR to connect to all BGP peers the VPC owner can access.

> - If the VPC owner has dedicated BGP peers, and account setting "use.system.bgp.peers" is set to false, configure BGP sessions in the VOC VR to connect to all dedicated BGP peers of the domain and the VPC owner.

Author: thanks, fixed
